staker-rewards/inspector/scripts/lib/cache.js

```diff
@@ -0,0 +1,212 @@
+    keepCreatedEvents.forEach((event) => {
+      keepsOnChain.push({
+        address: event.returnValues.keepAddress,
+        members: event.returnValues.members,
+        creationBlock: event.blockNumber,
+      })
+    })
+
+    const cachedKeepsCount = this.cache.get("keeps").value().length
+
+    console.log(
+        `Number of keeps created since block ` +
+        `[${previousRefreshBlock}]: ${keepsOnChain.length}`
+    )
+
+    console.log(`Number of keeps in the cache: ${cachedKeepsCount}`)
+
+    const actions = []
+    keepsOnChain.forEach((keep) => {
+      const currentlyCached = this.cache
+        .get("keeps")
+        .find({ address: keep.address })
+        .value()
+
+      if (!currentlyCached) {
+        actions.push(() => this.fetchKeep(keep))
+      }
+    })
+
+    if (actions.length === 0) {
+      console.log("Cached keeps list is up to date")
+    } else {
+      console.log(
+          `Fetching information about [${actions.length}] new keeps...`
+      )
+
+      const results = await pAll(actions, {
+        concurrency: CONCURRENCY_LIMIT,
+      })
+
+      console.log(
+          `Successfully fetched information about [${results.length}] new keeps`
+      )
+    }
+
+    const latestBlockNumber = keepCreatedEvents.slice(-1)[0].blockNumber
+    this.cache.assign({ lastRefreshBlock: latestBlockNumber }).write()
+
+    await this.refreshActiveKeeps()
+  }
+
+  async fetchKeep(keepData) {
+    return new Promise(async (resolve, reject) => {
+      try {
+        const { address, members, creationBlock } = keepData
+
+        this.cache
+            .get("keeps")
+            .push({
+              address: address,
+              members: members,
+              creationBlock: creationBlock,
+              status: await this.getKeepStatus(keepData)
+            })
+            .write()
+
+        console.log(
+            `Successfully fetched information about keep ${address}`
+        )
+
+        return resolve()
+      } catch (err) {
+        return reject(err)
+      }
+    })
+  }
+
+  async refreshActiveKeeps() {
+    const activeKeeps = this.getKeeps("active")
+
+    console.log(`Refreshing [${activeKeeps.length}] active keeps in the cache`)
+
+    const actions = []
+    activeKeeps.forEach(keepData => {
+      actions.push(() => this.refreshKeepStatus(keepData))
+    })
+
+    await pAll(actions, {concurrency: CONCURRENCY_LIMIT})
+  }
+
+  async refreshKeepStatus(keepData) {
+    console.log(`Checking current status of keep ${keepData.address}`)
+
+    const lastStatus = keepData.status
+    const currentStatus = await this.getKeepStatus(keepData)
+
+    if (lastStatus.name !== currentStatus.name) {
+      console.log(`Updating current status of keep ${keepData.address}`)
+
+      this.cache
+          .get("keeps")
+          .find({ address: keepData.address })
+          .assign({ status: currentStatus })
+          .write()
+    }
+  }
+
+  async getKeepStatus(keepData) {
```

Comment:
Let's document what's the format of the data returned by this function and especially what are timestamp values for closed/terminated/active cases as the way the first two work are a little different than how the third one work, at least to my understanding.

Reply:
I've added a comment. To clarify, the timestamp always means the same. It's the moment when the status has been set. Event for the `active` case despite it's calculated differently. The timestamp of the keep's creation block is actually the moment when the keep becomes `active`.